src/magnolia.hs

```diff
@@ -18,9 +17,37 @@ import PPrint
 import Syntax
 import Util
 
-type TopEnv = TcGlobalEnv
-data Status = Failure | Success -- TODO: move and expand statuses
-data CompilerMode = ReplMode | BuildMode String
+type TopEnv = Env (MPackage PhCheck)
+data CompilerMode = ReplMode | BuildMode FilePath | TestMode TestConfig FilePath
+
+data TestConfig = TestConfig { _testPass :: Pass
+                             , _testBackend :: Backend
+                             }
+
+-- | Compiler passes
+data Pass = CheckPass
+          | CodegenPass
+          | DepAnalPass
+          | ParsePass
+
+-- | Runs the compiler up to the pass specified in the test configuration, and
+-- logs the errors encountered to standard output.
+runTest :: TestConfig -> FilePath -> IO ()
+runTest config filePath = case _testPass config of
+  DepAnalPass -> runAndLogErrs $ depAnalPass filePath
+  ParsePass -> runAndLogErrs $ depAnalPass filePath >>= parsePass
+  CheckPass -> runAndLogErrs $ depAnalPass filePath >>= parsePass >>= checkPass
+  CodegenPass -> error "codegen not yet implemented"
+
+-- === debugging utils ===
+
+logErrs :: S.Set Err -> IO ()
+logErrs errs = pprintList (L.sort (toList errs))
```

Comment:
Wouldn't we lose information about when the error occurred if the list gets sorted before it is printed?

Reply:
Well, `errs` is a set to begin with – which makes no promises about conserving ordering of inputs (`OSet` would do that, but it didn't seem warranted), so this at least doesn't apply here.

From a more general standpoint, I don't think that there is much of a point from the point of view of the Magnolia developer to know in what order the errors occurred (?).

Another issue with not sorting is that topological sorts of our list of packages/modules are not unique, which means that we are not guaranteed to get a stable output if we do not sort the errors before printing them out. This is a problem, at the very least when it comes to our testing.

What do you think?

Reply:
Alright that clears it up for me, I was just curious about the reason for sorting in the first place. On second thought, I guess we don't really get anything useful information out of knowing the order. Thanks!